### PR TITLE
[codex] Harden ODELIA swarm runs for 8-site training

### LINKED ISSUE #314: Throttle per-round aggregated prediction export in swarm validate_and_train
## Context

The PR301 MST validation run on the 3-node startup kit was about 3.3x slower than the Apr 26 baseline (median round interval ~34 min vs ~10 min). The bottleneck is the per-round aggregated prediction export performed inside `validate_and_train()` in [application/jobs/_shared/custom/threedcnn_ptl.py:587-608](application/jobs/_shared/custom/threedcnn_ptl.py#L587-L608).

Each round, after validation, the swarm code calls `output_GT_and_classprobs_csv()` ([threedcnn_ptl.py:429-465](application/jobs/_shared/custom/threedcnn_ptl.py#L429-L465)) twice — once over the full train dataloader and once over the full val dataloader — running per-sample inference (`batch_size=1`) and appending rows to:

- `aggregated_model_gt_and_classprob_train.csv`
- `aggregated_model_gt_and_classprob_validation.csv`

In swarm mode this is enabled by default (see [threedcnn_ptl.py:63](application/jobs/_shared/custom/threedcnn_ptl.py#L63)); preflight/local mode already passes `output_GT_and_classprob=False` ([threedcnn_ptl.py:66](application/jobs/_shared/custom/threedcnn_ptl.py#L66)).

Cost scales linearly with `|train| + |val|` and with the slowest node, and currently runs every single round.

## Why it matters before scaling 3 → 8 nodes

Round time is governed by the slowest collaborator. With 8 sites and larger datasets, full-dataset inference + CSV writing every round will dominate wall-clock time and won't be informative more than once or twice.

## Proposal

Make per-round aggregated prediction export configurable in swarm mode:

- Default to **final-round only** (or every Nth round), not every round.
- Expose via env var (e.g. `ODELIA_PRED_EXPORT_EVERY=N`, `0` = final round only) and threaded through to `validate_and_train()`.
- Keep current behavior available for debugging.
- Confirm `GT_PredProb_Output_Callback` (the per-epoch local callback at [threedcnn_ptl.py:476-481](application/jobs/_shared/custom/threedcnn_ptl.py#L476-L481)) is left untouched or also gated — it currently fires every local epoch and contributes to the same CSV write pressure.

## Acceptance

- Configurable cadence with documented default.
- Round-interval regression closes most of the ~24 min gap on the 3-node MST run.
- 8-node projection is within tolerance for the planned full-scale experiment.

### PR DESCRIPTION
**Pull request by Ultimate-Storm**
## Summary

Harden ODELIA swarm runtime defaults for production-like 8-site training and include a small test-only pytest isolation fix.

This PR applies the hardening across all six ODELIA jobs: MST plus the five challenge jobs.

## Root Cause From The Failed 2-Round 1DC Attempts

The 2-round 1DivideAndConquer smoke surfaced several independent fragility points:

- The old `max_status_report_interval = 300` was unsafe for large challenge models; long validation/export or slow nodes could look dead even when training was still progressing.
- Swarm validation exported aggregated GT/prob CSVs by default, adding expensive post-round work that is not needed for normal swarm runs.
- Stale `live_sync.sh` daemons could keep syncing old startup-kit state into newer runs.
- DNS routing could drift between Tailscale and LAN routes; USZ/MHA-style routes need explicit per-site overrides.
- Global-best selection used `accuracy`, but the Lightning trainer emits `val/ACC`, so the selector metric did not match the actual model metric.

## Changes

- Raised ODELIA NVFlare server/client timeouts for MST + all challenge jobs:
  - Server: `start_task_timeout=3600`, `configure_task_timeout=1800`, `progress_timeout=43200`, `max_status_report_interval=7200`.
  - Client: `learn_task_timeout=43200`, `learn_task_ack_timeout=7200`, `final_result_ack_timeout=7200`, `peer_read_timeout=7200`, `last_result_transfer_timeout=7200`, `external_pre_init_timeout=1800`, `heartbeat_timeout=1800`, `wait_time_after_min_resps_received=1800`.
  - Kept `learn_task_abort_timeout=300` short.
- Changed all ODELIA `IntimeModelSelector.key_metric` values from `accuracy` to `val/ACC`.
- Disabled expensive swarm GT/prob export by default. It is now opt-in with `ODELIA_SWARM_EXPORT_GT_PROBS=1`.
- Hardened startup-kit live monitor sync:
  - Stop stale same-startup-dir `live_sync.sh` daemons before starting a new sync.
  - Remove stale PID files.
  - Keep final sync behavior, but guard it to run once.
  - Monitor detached swarm clients and stop swarm live-sync after normal client exit.
- Hardened deploy runner behavior:
  - Clean stale remote live-sync daemons before launching clients.
  - Add per-site `<SITE>_SERVER_IP_OVERRIDE` for explicit DNS routing.
  - Enable deploy-runner preprocess cache by default with per-site opt-out/dir override.
- Updated timeout and deploy runbook docs.
- Added a test-only fix in `tests/unit_tests/test_env_config.py` so the env-config unit tests mock `torch.multiprocessing.cpu_count` through `sys.modules` instead of forcing real PyTorch re-imports across tests.

## Validation

Static / unit checks:

- `bash -n scripts/deploy/run_deploy_test.sh` ✅
- `bash -n kit_live_sync/live_sync.sh` ✅
- Extracted generated client shell block from `docker_config/master_template.yml` and ran `bash -n` ✅
- `python3 -m py_compile application/jobs/_shared/custom/main.py` ✅
- `.venv/bin/pytest -q tests/unit_tests/test_env_config.py` ✅ (`21 passed`)
- `! rg "max_status_report_interval = 300|key_metric = \"accuracy\"" application/jobs/{ODELIA_ternary_classification,challenge_*}/app/config` ✅
- `rg "ODELIA_SWARM_EXPORT_GT_PROBS|output_GT_and_classprob=False|Swarm aggregated GT/prob export" application/jobs/_shared/custom/main.py` ✅
- `git diff --check` ✅

Startup-kit build:

- `./scripts/build/buildDockerImageAndStartupKits.sh -p application/provision/project_Odelia_allsites.yml --use-docker-cache` ✅
- Built image locally: `jefftud/odelia:1.4.5-dev.260610.702c30e` ✅
- Generated all site kits under `workspace/odelia_1.4.5-dev.260610.702c30e_allsites_test/prod_00` ✅
- Inspected generated `USZ_1/startup/docker.sh`: contains live-sync cleanup, swarm-client exit monitor, and preprocess-cache env forwarding ✅

Docker image was **not pushed** as part of this PR.

## Issue Handling

Closes #314

Refs #316: partially addresses timeout scaling, DNS routing, live-sync cleanup, and export bottleneck. Full 8-site operational validation remains follow-up.

Refs #160: partially addresses deploy/live-sync collection robustness, but does not implement full NVFlare-native log collection.

Refs #315: left open; ODELIA worker tuning still needs per-node measurement rather than a blanket default.

Refs #313: left open; full 1DC validation is an operational validation task after this hardening lands.

### CODE REVIEW

**Comment by Ultimate-Storm** [comment]
Validation update after the USZ+MHA 2-round 1DC smoke:

- Built and pushed `jefftud/odelia:1.4.5-dev.260610.2cf6e82` (`sha256:0c70b548be7b8e6a2135a3b3861886449a5cb2dbbef8dd66b335846add1a7f67`).
- Rebuilt all-sites startup kits with `--num-rounds 2`; kits are under `workspace/odelia_1.4.5-dev.260610.2cf6e82_allsites_test/prod_00/`.
- Earlier 2-round USZ+MHA 1DC smoke passed training + checkpoint collection in 4942s. Results: `workspace/deploy_test_results/usz_mha_1dc_2round_20260610T092600Z/`.
- Both `FL_global_model.pt` and `best_FL_global_model.pt` were collected from `MHA_1` and `USZ_1`; checksums match across sites for each artifact kind.
- Global-best selection using `val/ACC` was exercised in round 1 (`best metric=0.5841051042079926`).
- No `didn't report status for 300 seconds` failures, and swarm GT/prob export was disabled by default.
- Follow-up fix included after the smoke: live sync now writes the final finished heartbeat to both `heartbeat_final.json` and `heartbeat.json`, so the monitor current-state view should not remain stuck on `running` after normal stop.

Additional generated-image checks on the final tag:
- `challenge_1DivideAndConquer` has `num_rounds = 2`, raised timeouts, and `key_metric = \"val/ACC\"`.
- No ODELIA `max_status_report_interval = 300` or `key_metric = \"accuracy\"` matches remain.
- `ODELIA_SWARM_EXPORT_GT_PROBS` opt-in path is present and default swarm export remains off.

**Comment by Ultimate-Storm** [comment]
### Added commit `9960529`: raise swarm timeouts so a slow site can't abort the run

Recovered the real cause of the repeated `FINISHED:EXECUTION_EXCEPTION` on `challenge_1DivideAndConquer` from the server job log:

```
FATAL_SYSTEM_ERROR: client UKA_1 didn't report status for 7200 seconds
... failed to end workflow swarm_controller: only 4/6 clients responded, need 6
```

UKA_1's single round runs **>3 h** (~20× the steps/epoch of the others, which finish in ~15–30 min), so it blew past the **2 h** `max_status_report_interval` and the server killed the whole job.

Bumped (intentionally generous, tune down later) on all challenge models + `ODELIA_ternary_classification`:

| Param | Before | After |
|---|---|---|
| `max_status_report_interval` | 7200 (2h) | 86400 (24h) |
| `progress_timeout` | 43200 (12h) | 86400 (24h) |
| `learn_task_timeout` | 43200 (12h) | 86400 (24h) |
| `wait_time_after_min_resps_received` | 1800 (30m) | 21600 (6h) |

Streaming/ack timeouts and smoke jobs (cifar10/minimal_training/STAMP) untouched. Does not by itself make UKA keep pace — reducing UKA's per-round work is a separate follow-up.

**Comment by Ultimate-Storm** [comment]
Superseded by #345, which folds in these 24 h timeouts and adds wait-for-all sync (`min_responses_required`), automatic preflight checks, the cgroupfs host-fix helper, and the failure-mode runbook + guide updates. Closing in favour of #345.
